agent/src/util/device_plugin_service.rs

```diff
@@ -403,100 +809,337 @@ impl InstanceDevicePlugin {
-/// and false if it is already reserved.
-/// # More details
-/// Cases based on the usage slot (`device_usage_id`) value
-/// 1. device_usage\[id\] == "" ... this means that the device is available for use
-///     * (ACTION) return true
-/// 2. device_usage\[id\] == self.nodeName ... this means THIS node previously used id, but the DevicePluginManager knows that this is no longer true
-///     * (ACTION) return false
-/// 3. device_usage\[id\] == (some other node) ... this means that we believe this device is in use by another node and should be marked unhealthy
-///     * (ACTION) return error
-/// 4. No corresponding id found ... this is an unknown error condition (BAD)
-///     * (ACTION) return error
-fn slot_available_to_reserve(
-    device_usage_id: &str,
+pub async fn get_instance_device_usage_state(
     node_name: &str,
-    instance: &InstanceSpec,
-) -> Result<bool, Status> {
-    if let Some(allocated_node) = instance.device_usage.get(device_usage_id) {
-        if allocated_node.is_empty() {
-            Ok(true)
-        } else if allocated_node == node_name {
-            Ok(false)
-        } else {
-            trace!("slot_available_to_reserve - request for device slot {} previously claimed by a diff node {} than this one {} ... indicates the device on THIS node must be marked unhealthy, invoking ListAndWatch ... returning failure, next scheduling should succeed!", device_usage_id, allocated_node, node_name);
-            Err(Status::new(
-                Code::Unknown,
-                "Requested device already in use",
-            ))
+    instance_name: &str,
+    instance_namespace: &str,
+    capacity: &i32,
+    kube_interface: Arc<impl KubeInterface>,
+) -> Vec<(String, DeviceUsageStatus)> {
+    let mut device_usage_states = Vec::new();
+    match kube_interface
+        .find_instance(instance_name, instance_namespace)
+        .await
+    {
+        Ok(kube_akri_instance) => {
+            for (device_name, device_usage_string) in kube_akri_instance.spec.device_usage {
+                let device_usage_status = match DeviceUsage::from_str(&device_usage_string) {
+                    Ok(device_usage) => get_device_usage_state(&device_usage, node_name),
+                    Err(_) => DeviceUsageStatus::Unknown,
+                };
+                device_usage_states.push((device_name.clone(), device_usage_status));
+            }
+            device_usage_states
         }
-    } else {
-        // No corresponding id found
-        trace!(
-            "slot_available_to_reserve - could not find {} id in device_usage",
-            device_usage_id
-        );
-        Err(Status::new(
-            Code::Unknown,
-            "Could not find device usage slot",
-        ))
+        Err(_) => (0..*capacity)
+            .map(|x| {
+                (
+                    format!("{}-{}", instance_name, x),
+                    DeviceUsageStatus::Unknown,
+                )
+            })
+            .collect(),
```

Comment:
would it be easier to surface possible bugs if we bubbled up an error here instead of DeviceUsageStatus::Unknown?  Is there a valid state where Unknown is expected?

Reply:
Unknown is a valid state, mostly due to the akrii is deleted and the Agent hasn't update instance map

Reply:
thanks!  sorry for repeating the Unknown questions in this PR and the other one.  are both PRs going to be merged or is this the one PR that will be merged?

Reply:
(just trying to figure out where best to focus energy)

Reply:
since the change scope of this PR is too big, I extracted changes for device usage to a separate PR.  Let's focus on the device usage PR first and review this one after the device usage PR committed.
https://github.com/project-akri/akri/pull/628